homeassistant/components/overseerr/__init__.py

```diff
@@ -27,3 +46,61 @@ async def async_setup_entry(hass: HomeAssistant, entry: OverseerrConfigEntry) ->
 async def async_unload_entry(hass: HomeAssistant, entry: OverseerrConfigEntry) -> bool:
     """Unload a config entry."""
     return await hass.config_entries.async_unload_platforms(entry, PLATFORMS)
+
+
+class OverseerrWebhookManager:
+    """Overseerr webhook manager."""
+
+    def __init__(self, hass: HomeAssistant, entry: OverseerrConfigEntry) -> None:
+        """Initialize Overseerr webhook manager."""
+        self.hass = hass
+        self.entry = entry
+        self.client = entry.runtime_data.client
+
+    async def register_webhook(self) -> None:
+        """Register webhook."""
+        async_register(
+            self.hass,
+            DOMAIN,
+            self.entry.title,
+            self.entry.data[CONF_WEBHOOK_ID],
+            self.handle_webhook,
+            allowed_methods=[METH_POST],
+        )
+        url = async_generate_url(self.hass, self.entry.data[CONF_WEBHOOK_ID])
+        if not await self.check_need_change(url):
+            return
+        LOGGER.debug("Setting Overseerr webhook to %s", url)
+        if not await self.client.test_webhook_notification_config(url, JSON_PAYLOAD):
+            LOGGER.debug("Failed to set Overseerr webhook")
+            return
+        await self.client.set_webhook_notification_config(
+            enabled=True,
+            types=REGISTERED_NOTIFICATIONS,
+            webhook_url=url,
+            json_payload=JSON_PAYLOAD,
+        )
+
+    async def check_need_change(self, url: str) -> bool:
+        """Check if webhook needs to be changed."""
+        current_config = await self.client.get_webhook_notification_config()
+        return (
+            not current_config.enabled
+            or current_config.options.webhook_url != url
+            or current_config.options.json_payload != json.loads(JSON_PAYLOAD)
+            or current_config.types != REGISTERED_NOTIFICATIONS
+        )
+
+    async def handle_webhook(
+        self, hass: HomeAssistant, webhook_id: str, request: Request
+    ) -> Response:
+        """Handle webhook."""
+        data = await request.json()
+        LOGGER.debug("Received webhook payload: %s", data)
+        if data["notification_type"].startswith("MEDIA"):
+            await self.entry.runtime_data.async_refresh()
```

Comment:
why are we refreshing the data here?

Reply:
Because this data is about the made requests, so when we get this webhook we know we can update the data. In the future I will add event entities, but that would make this PR a lot bigger, so this was a neat way of adding this